<div style="border:solid Chocolate 2px; padding: 40px">

<b> Александр, привет!👋</b>

Меня зовут Алексей Гриб, и я буду ревьюером твоего проекта. 

Сразу хочу предложить в дальнейшем общаться на "ты" - надеюсь, так будет комфортнее:) Но если это неудобно, обязательно дай знать, и мы придумаем что-нибудь ещё!
    
Цель ревью - не искать ошибки в твоём проекте, а помочь тебе сделать твою работу ещё лучше, устранив недочёты и приблизив её к реальным задачам специалиста по работе с данными. Поэтому не расстраивайся, если что-то не получилось с первого раза - это нормально, и это поможет тебе вырасти!
    
Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> похвала, рекомендации «со звёздочкой», полезные лайфхаки, которые сделают и без того красивое решение ещё более элегантным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> некритичные ошибки или развивающие рекомендации на будущее. 
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Критичные ошибки, которые обязательно нужно исправить.
</div>

    
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 
    
Ты также можешь задавать свои вопросы, реагировать на мои комментарии, делать пометки и пояснения - полная творческая свобода! Но маленькая просьба - пускай они будут отличаться от моих комментариев, это поможет избежать путаницы в нашем общении:)
Например, вот так:
    
<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*твой текст*
</div>
    
Давай посмотрим на твой проект!

# Проектная работа: Анализ сервиса для чтения книг по подписке.

Цель исследования: 

Проанализировать базу данных сервиса для чтения книг по подписке и сформулировать ценностное предложение для нового продукта.

# Введение

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша задача проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**План иследования:**

Шаг 1. Загрузка данных.

Шаг 2. Исследование таблиц.

Шаг 3. Анализ таблиц.

Шаг 4. Вывод

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Хорошее вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!

## Шаг 1. Загрузка данных.

In [1]:
import pandas as pd
import sqlalchemy as sa
from sqlalchemy import create_engine 

In [2]:
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}

In [3]:
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [4]:
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

In [5]:
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали, коннектор сделали - отлично!

## Исследование таблиц:

### Таблица `books`

In [6]:
books =( '''
SELECT 
    *
FROM books
    LIMIT 5
''')    
get_sql_data(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [7]:
books =( '''
SELECT 
    COUNT(book_id),
    COUNT(DISTINCT(book_id)) distinct,
    COUNT(DISTINCT(title)) distinct_book
FROM books
    LIMIT 5
''')    
get_sql_data(books)

count  distinct  distinct_book
0   1000      1000            999

Данные в таблице `books` представлены следующими столбцами: `book_id` идентификатор книги, `author_id` идентификатор автора,  `title` название книги, `num_pages` количество страниц, `publication_date` дата публикации книги, `publisher_id` идентификатор издателя. В таблице 1000 строк. Уникальных названий книг 999

### Таблица `authors`

In [8]:
authors =( '''
SELECT 
    *
FROM authors
    LIMIT 5
''')    
get_sql_data(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
authors =( '''
SELECT 
    COUNT(author_id),
    COUNT(DISTINCT(author_id)) distinct,
    COUNT(DISTINCT(author)) distinct_author
FROM authors
    LIMIT 5
''')    
get_sql_data(authors)

count  distinct  distinct_author
0    636       636              636

Данные в таблице `authors` представлены следующими столбцами: `author_id` идентификатор автора,  `author` имя автора.  В таблице 636 строк. Уникальных авторов 636

### Таблица `publishers`

In [10]:
publishers =( '''
SELECT 
    *
FROM publishers
    LIMIT 5
''')    
get_sql_data(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
publishers =( '''
SELECT 
    COUNT(publisher_id),
    COUNT(DISTINCT(publisher_id)) distinct,
    COUNT(DISTINCT(publisher)) distinct_publisher
FROM publishers
    LIMIT 5
''')    
get_sql_data(publishers)

count  distinct  distinct_publisher
0    340       340                 340

Данные в таблице `publishers` представлены следующими столбцами: `publisher_id` идентификатор издателя, `publisher` название издательства. В таблице 340 строк. Уникальных издательств 340

### Таблица `ratings`

In [12]:
ratings =( '''
SELECT 
    *
FROM ratings
    LIMIT 5
''')    
get_sql_data(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
ratings =( '''
SELECT 
    COUNT(rating_id),
    COUNT(DISTINCT(rating_id)) distinct,
    COUNT(DISTINCT(username)) distinct
FROM ratings
    LIMIT 5
''')    
get_sql_data(ratings)

count  distinct  distinct
0   6456      6456       160

Данные в таблице `ratings` представлены следующими столбцами: `rating_id` идентификатор оценки, `book_id` идентификатор книги,   `username` имя пользователя, оставившего оценку, `rating` оценка книги.  В таблице 6456 строк. Уникальных оценевающих пользователей 160.

### Таблица `reviews`

In [14]:
reviews =( '''
SELECT 
    *
FROM reviews
    LIMIT 5
''')    
get_sql_data(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [15]:
reviews =( '''
SELECT 
    COUNT(review_id),
    COUNT(DISTINCT(review_id)) distinct,
    COUNT(DISTINCT(username)) distinct
FROM reviews
    LIMIT 5
''')    
get_sql_data(reviews)

count  distinct  distinct
0   2793      2793       160

Данные в таблице `reviews` представлены следующими столбцами: `review_id` идентификатор обзора, `book_id` идентификатор книги,   `username` имя пользователя написавшего обзор, `text` текст обзора.  В таблице 2793 строк. Уникальных критикующих пользователей 160.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Данные осмотрели - отлично!

## Детализированное иследование

### Анализ даты публикации книги.
Посчитайте, сколько книг вышло после 1 января 2000 года

In [16]:
zad_1 =( '''

SELECT 
    COUNT(book_id) 
FROM books
    WHERE CAST(publication_date AS date) > '2000-01-01'

''')    
get_sql_data(zad_1)

count
0    819

Начиная с первого января 2000 года, на сервиве по подписке книг, было опубликованно 819 книг 

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> При фильтрации использован неверный год.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*Изменил год на 2000*
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

### Анализ количества обзоров и средней оценки.

Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [17]:
zad_2 =( '''

WITH 

rat AS (
SELECT 
    a.author book,
    AVG(rt.rating) avg_rating
FROM books AS b 
JOIN authors a ON b.author_id=a.author_id
LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
    GROUP BY a.author
    ORDER BY a.author
),

rev AS(
SELECT 
    a.author book,
    COUNT(rv.review_id) count_review
FROM books AS b 
JOIN authors a ON b.author_id=a.author_id
LEFT JOIN reviews AS rv ON b.book_id=rv.book_id
    GROUP BY a.author
    ORDER BY a.author
)

SELECT 
    rat.book,
    rev.count_review,
    ROUND(rat.avg_rating, 2) avg_rating
FROM rat 
JOIN rev ON rat.book=rev.book
    LIMIT 5
''')    
get_sql_data(zad_2)

book  count_review  avg_rating
0                      A.S. Byatt             3        4.00
1  Aesop/Laura Harris/Laura Gibbs             2        3.75
2                 Agatha Christie            28        4.28
3                   Alan Brennert             4        3.00
4        Alan Moore/David   Lloyd             3        4.22

Пользовательские рейтенги книг как и количество проставленных оценок отличаются у разных книг.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Метрики для книг посчитаны верно! Здорово, что метрики посчитаны и для тех книг, для которых нет отзывов - довольно частая ошибка в этом задании:)

</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Для идентификации книг вместо `id` стоит вывести явное название.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*Поменял индентификатор книг*
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

In [18]:
zad_2 =( '''

WITH 

rat AS (
SELECT 
    b.book_id book,
    AVG(rt.rating) avg_rating
FROM books AS b 
LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
    GROUP BY b.book_id
    ORDER BY b.book_id
),

rev AS(
SELECT 
    b.book_id book,
    COUNT(rv.review_id) count_review
FROM books AS b 
LEFT JOIN reviews AS rv ON b.book_id=rv.book_id
    GROUP BY b.book_id
    ORDER BY b.book_id
),

alls AS(
SELECT 
    rat.book,
    rev.count_review,
    ROUND(rat.avg_rating, 2) avg_rating
FROM rat 
JOIN rev ON rat.book=rev.book
)
 
SELECT 
    ROUND(AVG(count_review), 2) avg_review,    
    MAX(count_review) max_review,
    MIN(count_review) min_review,
    ROUND(AVG(avg_rating), 2) avg_rating,
    MAX(avg_rating) max_rating,
    MIN(avg_rating) min_rating
FROM
    alls

    
   
''')    
get_sql_data(zad_2)

avg_review  max_review  min_review  avg_rating  max_rating  min_rating
0        2.79           7           0         3.9         5.0         1.5

Среднее количество просмотров рамно 2,79, максимальное количество просмотров 7,
минимальное 0. Средний рейтинг книги равен 3,9 максимальный 5, минимальный 1,5

### Анализ издательств

Определите издательство, которое выпустило наибольшее число книг толще
50 страниц — так вы исключите из анализа брошюры;

In [19]:
zad_3 =( '''

SELECT 
    COUNT(b.book_id),
    p.publisher_id,
    publisher
FROM books b
LEFT JOIN publishers p ON b.publisher_id=p.publisher_id
    WHERE num_pages > 50
    GROUP BY p.publisher_id
    ORDER BY COUNT(b.book_id) DESC
LIMIT 1

''')    
get_sql_data(zad_3)

count  publisher_id      publisher
0     42           212  Penguin Books

`Penguin Books` Это издательство выпустило 42 книги объемом более 50 страниц, что является максимальным количеством таких книг 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Издательство опредили верно! Здорово, что дополнительно выведено количество книг издательства:)

</div>

### Анализ авторов

Определите автора с самой высокой средней оценкой книг — учитывайте
только книги с 50 и более оценками;

In [20]:
zad_4 =( '''
SELECT 
    a.author_id
FROM author a
LEFT JOIN books b ON a.author_id=b.author_id
WHERE b.book_id IN (
    SELECT 
        b.book_id
    FROM books b
    JOIN authors a ON b.author_id=a.author_id
    LEFT JOIN ratings r ON b.book_id=r.book_id
        GROUP BY b.book_id
        HAVING COUNT(r.username) >= 50
        ORDER BY AVG(r.rating) DESC, COUNT(r.username) DESC
   )
LIMIT 1
''')    
get_sql_data(zad_4)

author_id
0        236

In [21]:
zad_4 =( '''

SELECT 
    a.author,
    ROUND(AVG(rating), 6) avg_rating
FROM authors a
LEFT JOIN books b ON a.author_id=b.author_id
LEFT JOIN ratings r ON r.book_id=b.book_id
WHERE b.book_id IN (
    SELECT 
        b.book_id
    FROM books b
    JOIN authors a ON b.author_id=a.author_id
    LEFT JOIN ratings r ON b.book_id=r.book_id
        GROUP BY b.book_id
        HAVING COUNT(r.username) >= 50
        ORDER BY  COUNT(r.username) DESC)
GROUP BY a.author    
ORDER BY AVG(rating) desc
LIMIT 1


''')    
get_sql_data(zad_4)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Максимальный средний рейтинг, среди книг с пятидесятью и более оценками, равен 4,29 и он принаджежит слудующему автору `J.K. Rowling/Mary GrandPré`. 

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> В запросе выполняется фильрация по количеству страниц, а не по количеству оценок.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*Поменял фильтрацию*
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.2. <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Так как в первом запросе данные сгруппированы на уровне автора, то фильтр `HAVING`, призванный отбросить книги с количеством отзывов менее 50, перестаёт работать: у каждого автора может быть по несколько книг с количеством отзывов менее 50, но если просуммировать отзывы по всем книгам в пределах автора (а именно так и происходит при группировке на уровне автора), то 50 и более отзывов легко набирается. 
    
Корректнее здесь либо группировать сначала на уровне книги, тогда фильтр `HAVING` будет работать, либо оставить группировку по автору, но в подзапросе найти книги, которые могут участвовать в расчёте рейтинга - тогда фильтр `HAVNIG` будет в этом подзапросе.
    
Также нужно вывести там рейтинг автора, а не только автора.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.3 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Сейчас данные сгруппированы на уровне книги, и в запросе вывден не средний рейтинг автора, а средний рейтинг его самой популярной книги. Чтобы при текущей структуре запроса получить средний рейтинг автора, нужно взять среднее из средних рейтингов его книг.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*пределал*
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.4 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Фильтрацию нужно выполнять не по автору книги, а по самим книгам, для которых выполняется условие `>= 50`.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*Да, я еще вчера вечером сделал и этот вариант запроса, но спешил и был не внимателен.*
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.5 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Теперь верно.

### Анализ обзоров

Посчитайте среднее количество обзоров от пользователей, которые поставили
больше 48 оценок.

In [22]:
zad_5 =( '''

SELECT 
    AVG(count_reviews) avg_reviews
FROM (
    SELECT 
        COUNT(review_id) count_reviews,
        username
    FROM reviews
        WHERE username IN (
            SELECT 
                username
            FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 48
            )
        GROUP BY username    
) AS f


''')    
get_sql_data(zad_5)

avg_reviews
0         24.0

In [23]:
zad_5 =( '''

SELECT 
    COUNT(count_reviews) count_reviews
FROM (
    SELECT 
        COUNT(review_id) count_reviews,
        username
    FROM reviews
        WHERE username IN (
            SELECT 
                username
            FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 48
            )
        GROUP BY username    
) AS f


''')    
get_sql_data(zad_5)

count_reviews
0             13

Средннее количество обзоров пользователей, которые поставили более 48 оценок, равно 24 обзора. Общее количество таких пользователей 13

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, посчитано верно!

## Вывод

Данные в таблице `books` представлены следующими столбцами: `book_id` идентификатор книги, `author_id` идентификатор автора,  `title` название книги, `num_pages` количество страниц, `publication_date` дата публикации книги, `publisher_id` идентификатор издателя. В таблице 1000 строк. Уникальных названий книг 999

Данные в таблице `authors` представлены следующими столбцами: `author_id` идентификатор автора,  `author` имя автора.  В таблице 636 строк. Уникальных авторов 636

Данные в таблице `publishers` представлены следующими столбцами: `publisher_id` идентификатор издателя, `publisher` название издательства. В таблице 340 строк. Уникальных издательств 340

Данные в таблице `ratings` представлены следующими столбцами: `rating_id` идентификатор оценки, `book_id` идентификатор книги,   `username` имя пользователя, оставившего оценку, `rating` оценка книги.  В таблице 6456 строк. Уникальных оценевающих пользователей 160.

Данные в таблице `reviews` представлены следующими столбцами: `review_id` идентификатор обзора, `book_id` идентификатор книги,   `username` имя пользователя написавшего обзор, `text` текст обзора.  В таблице 2793 строк. Уникальных критикующих пользователей 160.

Начиная с первого января 2000 года, на сервиве по подписке книг, было опубликованно 819 книг

Пользовательские рейтенги книг как и количество проставленных оценок отличаются у разных книг.

Среднее количество обзоров ровно 2,79, максимальное количество просмотров 7,
минимальное 0. Средний рейтинг книги равен 3,9 максимальный 5, минимальный 1,5

`Penguin Books` Это издательство выпустило 42 книги объемом более 50 страниц, что является максимальным количеством таких книг 

Максимальный средний рейтинг, среди книг с пятидесятью и более оценками, равен 4,29 и он принаджежит слудующему автору `J.K. Rowling/Mary GrandPré`. 


Средннее количество обзоров пользователей, которые поставили более 48 оценок, равно 24 обзора.

### Итоговый вывод

Куплен маленький сервис для чтения книг по подписке,  в сервисе представлено всего 1000 книг 81% из которых современные.  На сервисе публикуются 636 авторов из 340  издательств. Рейтинги и обзоры на сервисе отличаются, они "не накручены", но активная аудитория малеленькая, всего из 160 пользователей 13 активных.

### Рекомендации
- необходимо увеличивать количество книг на сервисе.
- необходимо привлекать как авторов так и издательства.
- Необходимо увеличивать как саму аудиторию читателей так и ее активность.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Отличный итоговый вывод по проекту!

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту.

Александр, хорошая работа над проектом - часть заданий у тебя выполнены верно!

Мне нравится твой аналитический подход к выполнению проекта, ты соблюдаешь структуру работы, выполняешь её последовательно - это очень хорошо! Шаги проекта выполнены по порядку согласно плану проекта, нет смысловых и структурных ям. Важно, что не забываешь про выводы.
    
SQL запросы написаны хорошо и оптимально, в них нет избыточных действий, они возвращают верный результат - это здорово, кажется, тебе удалось здорово разобраться в теме.
    
Есть небольшие рекомендации по улучшению проекта - некритичная мелочь, которая не займёт много времени, но сделает твой проект стилистически более совершенным. Будет классно, если эти рекомендации будут учтены в твоей работе:)
    
К нескольким заданиям стоит вернуться - к ним я оставил комментарии красного цвета с пояснениями внутри. Думаю, ты быстро разберёшься, в чём дело.

Также буду рад ответить на любые твои вопросы по проекту или на какие-либо другие, если они у тебя имеются - оставь их в комментариях, и я постараюсь ответить:)
    
Жду твой проект на повторном ревью. До встречи:)

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту.

Александр, вот мы и преодолели финишную черту - поздравляю с завершением финального спринта!
    
Проделана без сомнения крутая и тяжёлая работа, и все проекты финального спринта удадось выполнить на достойном уровне - уверен, если решишь развиваться дальше в направлении аналитики, тебя ждёт успех и хороший рост:)
    
Спасибо за хорошую работу над проектами финального спринта, желаю дальнейшего профессионального роста и развития, проектов для применения на практике полученных значений, а также адекватных заказчиков:)